src/span.js

```diff
@@ -0,0 +1,220 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import * as constants from './constants.js';
+import * as thrift from './thrift.js';
+import Utils from './util.js';
+import SpanContext from './span_context.js';
+
+export default class Span {
+    _tracer: any;
+    _name: string;
+    _spanContext: SpanContext;
+    _start: number;
+    _duration: number;
+    _firstInProcess: boolean;
+    _isClient: boolean;
+    _peer: Endpoint;
+    _annotations: Array<Annotation>;
+    _binaryAnnotations: Array<BinaryAnnotation>;
+    _baggageHeaderCache: any;
+
+    constructor(tracer: any,
+                name: string,
+                spanContext: SpanContext,
+                start: number = Utils.getTimestampMicros(),
+                firstInProcess: boolean = false,
+                annotations: Array<Annotation> = [],
+                binaryAnnotations: Array<BinaryAnnotation> = []
+    ) {
+        this._tracer = tracer;
+        this._name = name;
+        this._spanContext = spanContext;
+        this._start = start;
+        this._firstInProcess = firstInProcess;
+
+        // logs
+        this._annotations = annotations;
+
+        // tags
+        this._binaryAnnotations = binaryAnnotations;
+    }
+
+    static _getBaggageHeaderCache() {
+        if (!this._baggageHeaderCache) {
+            // $FlowIgnore - flow doesn't seem to like how statics are declared
+            this._baggageHeaderCache = {};
+        }
+
+        return this._baggageHeaderCache;
+    }
+
+    /**
+     * Returns a normalize key.
+     *
+     * @param {string} key - The key to be normalized for a particular baggage value.
+     * @return {string} - The normalized key (lower cased and underscores replaced, along with dashes.)
+     **/
+    _normalizeBaggageKey(key: string) {
+       // TODO(oibe) should be moved to span in newest opentracing version
+        let baggageHeaderCache = this.constructor._getBaggageHeaderCache();
+        if (key in baggageHeaderCache) {
+            return baggageHeaderCache[key];
+        }
+
+        // This stackoverflow claims this approach is slower than regex
+        // http://stackoverflow.com/questions/1144783/replacing-all-occurrences-of-a-string-in-javascript
```

Comment:
won't surprise me, you're creating an array of strings and then joining them into a new string. Just use regex. The whole point of your baggage header cache is that this is only done a few times